lib/ex_doc/application.ex

```diff
@@ -17,6 +17,11 @@ defmodule ExDoc.Application do
         match?("makeup_" <> _, Atom.to_string(app)),
         do: Application.ensure_all_started(app)
 
-    Supervisor.start_link([ExDoc.Refs], strategy: :one_for_one)
+    children = [
+      ExDoc.Refs,
+      ExDoc.WarningCounter
```

Comment:
Instead of introducing some global state, I wonder if you could create a counter and store it in the `config`. And then you call `warning(config, ...)` which will bump the counter. This means it is naturally async and you don't need the additional process. :)

Reply:
In the same vein, perhaps `mix docs` would return https://hexdocs.pm/mix/1.15/Mix.Task.Compiler.Diagnostic.html structs? Could be useful for IDEs? Or since ExDoc is not a _compiler_ it’s a bad idea?

Reply:
Yeah, I don't see how they could be used if it is not a compiler :(

Reply:
Passing config simplifies a lot the implementation.
The problem I am facing now is that there are functions that generate a warning and have no notion of config,
such as `ExDoc.Markdown.Earmark.to_ast/2`

Reply:
We would need to either:

1. pass an optional function to them to configure how logging happens
2. have those functions return their warnings so we can emit them

Reply:
I’d do the latter, return `{ast, %{warnings: warnings}}` or something like that.

I think such APIs that return instead of emit warnings are nice because we can test without capture_io and thus async. In this particular case since it’s not useful to return these warnings from `mix docs`, what we can also do is push all ExDoc warnings via some ExDoc.Utils.warn, inside call Application.put_env(:ex_doc, :warnings_emitted, true), check for this app env at the end of the run and call it a day. :)

Reply:
The counter will tell us if warnings were emitted or not. So having a config plus function return should be enough!

Reply:
I will work on this solution and let you know when it is done.

Reply:
Thank you ❤️